File: src/main/java/com/oycm/algorithm/aba/basic/Solution_10.scala
package com.oycm.algorithm.aba.basic

object Solution_10 {

  /**
   * 1004. 最大连续1的个数 III 1656
   * https://leetcode.cn/problems/max-consecutive-ones-iii/description/
   *
   * @param nums 二进制数组
   * @param k    翻转 k 个 0
   * @return 执行翻转后连续 1 的最大个数
   */
  def longestOnes(nums: Array[Int], k: Int): Int = {
    /*
    问题就是求 0 不超过 k 个的最长子数组
    */
    var ans = 0
    var l = 0
    var cnt = 0
    for (r <- nums.indices) {
      if (nums(r) == 0) {
        cnt += 1
      }
      while (cnt > k) {
        if (nums(l) == 0) {
          cnt -= 1
        }
        l += 1
      }
      ans = Math.max(ans, r - l + 1)
    }

    ans
  }
}

File: src/main/java/com/oycm/algorithm/c/Solution_10.java
package com.oycm.algorithm.c;


public class Solution_10 {

    /**
     * 3708. <a href="https://leetcode.cn/problems/longest-fibonacci-subarray/description/">最长斐波那契子数组</a> 1381
     * <p>
     * 斐波那契数组 ： nums[i] + nums[i+1] = nums[i+2]
     *
     * @param nums
     * @return 求 nums 中 最长的 斐波那契 数组长度
     */
    public int longestSubarray(int[] nums) {
        int ans = 0;
        int cnt = 2;
        int n = nums.length;
        for (int i = 2; i < n; i++) {
            if (nums[i - 2] + nums[i - 1] == nums[i]) {
                cnt++;
                ans = Math.max(cnt, ans);
            } else {
                cnt = 2;
            }
        }
        return ans;
    }

}
